### ISSUE TITLE
Error: panic: database is locked

### ISSUE THREAD

**Issue by NotoriousRebel** [g0ldencybersec/CloudRecon#2]
Hello, while runnning the tool like so: `time CloudRecon store -i ipv4_merged.txt -p 443 -t 10 -c 50`. I encountered this error 14 hours later... For context: [ipv4_merged.txt](https://github.com/lord-alfred/ipranges/blob/main/oracle/ipv4_merged.txt)

```panic: database is locked

goroutine 45 [running]:
main.runCloudStore.func1(0x0?)
        /root/go/pkg/mod/github.com/g0ldencybersec/!cloud!recon@v1.0.3/store.go:65 +0x468
created by main.runCloudStore in goroutine 1
        /root/go/pkg/mod/github.com/g0ldencybersec/!cloud!recon@v1.0.3/store.go:49 +0x17e

real    837m40.120s
user    1m50.644s
sys     1m45.807s
```

There is no debug or verbose mode so I can't provide anymore details besides this, I did encounter the same error with `-c 100` which is why I changed it to `-c 50` but to no avail it seems.

**Comment by NotoriousRebel** [comment]
Happened again but this one hurts a lot more:

```
panic: database is locked

goroutine 8 [running]:
main.runCloudStore.func1(0x0?)
        /root/go/pkg/mod/github.com/g0ldencybersec/!cloud!recon@v1.0.3/store.go:60 +0x485
created by main.runCloudStore in goroutine 1
        /root/go/pkg/mod/github.com/g0ldencybersec/!cloud!recon@v1.0.3/store.go:49 +0x17e

real    2778m27.455s
user    5m13.007s
sys     5m47.802s
```

**Comment by g0ldencybersec** [comment]
So the problem with store (that I am trying to find a change/way around) is that the DB object does not handle concurrent writes to the DB. So even though you can spin up goroutines. If two goroutines try to write to the DB at the same time. The first one will "lock" the DB to do a write. And the second one will error. That is what is happening to you. In the talk about the tool I mentioned that the store functionality is very limited due to that caveat.

Multiple connections (possibly owned by different processes) can read data from the same database at the same time, but only one can write.

I have a solution in mind I am trying to implement so that each goroutine you spin up isn't writing, but instead is pushing to a separate channel for a "writer" goroutine. That HOPEFULLY will solve this issue. But it is a work in progress.

So basically the tool was running until it happen to come across overlapping writes.

**Comment by NotoriousRebel** [comment]
Interesting, I suppose it's somewhat of an edge case as it depends on what is being written to the database at a given time. I am more of a proponent of an alternative solution raised in #1 as that would allow for more flexibility and entirely avoid this error as well. I am not too familar with Go but a mutex may or may not be needed; I have provided the exact command and where the IPs were derived from which should expedite testing and determine if a mutex is needed for file writing when dealing with multiple concurrent goroutines. Thanks for the awesome tool, although it may have crashed the sheer speed of it is nothing to scoff at, paired with the simplicity, just needs a bit more polish to get to the next level.

**Comment by mr-rizwan-syed** [comment]
cat ipv4_merged.txt | tlsx -json -silent | jq -c 'select(.probe_status) | { "ip": .ip, "port": .port, "organization_name": .issuer_org[0], "common_name": .subject_cn, "san": .subject_an[] }' 2>/dev/null 1> certdb.json
